First back-end

kolayne opened the pull request:
Hi! This is the first version of the project's back-end. I've been trying to reduce its size as far as possible, however, it turned out to be somewhat large, so probably it's a good idea not to review the whole PR in one sitting and review it by parts several times instead, but it's totally up to you.

So, what is in the PR?
- Updates of some meta files (added usage to README.md, created .gitignore and a config sample file)
- Some basic HTML templates (which are most likely going to be rewritten completely, but I've added them to make it easier to test/debug the app. They may also be useful for people, who are going to create the UI: they may want to see sample templates to understand, how to create theirs)
- SQL bridge ~~(in fact, most of the file was copied from another solution I've found on the internet. Note that it uses the `reflect` package of go, which might be hard to understand without reading some docs, so if you hadn't worked with it before, I **highly recommend** reading [the article](https://blog.golang.org/laws-of-reflection) by the developers of go, giving the introduction to the library)~~
- HTTP server with the most basic pages and one API endpoint (I am planning to add some documentation of my API in a separate PR to make it really easy for front-end developers to integrate with the back-end)

kolayne committed:
Added .gitignore

Generated with gitignore.io, tags: Go, Intellij+all, VisualStudioCode

kolayne committed:
Added first simple version of a website based on gorilla/mux

kolayne committed:
Minor improvements

- Added `panicIfError`
- Added requests methods restriction

kolayne committed:
Added /login page

kolayne committed:
Renamed main.go -> webhandlers.go and added `Usage` section to README

kolayne committed:
Added /logout page

kolayne committed:
Merge branch 'master' into first_backend

kolayne committed:
Added structures describing objects

kolayne committed:
Added connection with database via `sqlx`

kolayne committed:
Added first api endpoint (/api/days/brief)

kolayne committed:
Added ability to set cookie storage key via environment vars

DarkSquirrelComes commented on config/postgres_credentials_example.txt:
```diff
@@ -0,0 +1 @@
+user=USER password=PASSWORD dbname=DBNAME
```
Why you choose such config format?

kolayne replied:
That's the format accepted by the [`PostgreSQL` driver](https://github.com/lib/pq) we are using. Is there a reason to use a different format and then convert?

DarkSquirrelComes commented on src/database_bridge.go:
```diff
@@ -0,0 +1,147 @@
+// slice (in this case the queried data is appended to its end) or a struct (then only the first selected value is
+// written to it, others are suppressed)). If `result` is a pointer to a struct, but no results are returned by the
+// query, the appropriate error is returned, `result` is not updated
+func dbQuery(result interface{}, query string, args ...interface{}) error {
```
As i know, sql.Open(...) creates a connection pool and you shouldn't re-create it every time. Check it please.

kolayne replied:
You are right, I'll fix it

DarkSquirrelComes commented on src/common.go:
```diff
@@ -0,0 +1,10 @@
+//Panics if `err` is not `nil`
+//
+// @param err Error to be checked for equality to `nil`
+func panicIfError(err error) {
```
May be we should write some log here? And is it good error-handling pattern? (I really dont know. But i know a guy, who knows, i'll ask him)

kolayne replied:
I agree with logging. But I don't know about good patterns of error-handling patterns, it's just an idea a came up with, and this function was one of the reasons to request your review

DarkSquirrelComes commented on src/webhandlers.go:
```diff
@@ -0,0 +1,170 @@
+	js, err := json.Marshal(days)
+	panicIfError(err)
+	w.Header().Set("Content-Type", "application/json")
+	_, _ = w.Write(js)
```
_, _

kolayne replied:
Oops

DarkSquirrelComes commented on src/webhandlers.go:
```diff
@@ -0,0 +1,170 @@
+	http.Redirect(w, r, "..", 302)
+}
+
+func handleApiDaysBrief(w http.ResponseWriter, r *http.Request) {
```
Is size of this function ok?

kolayne replied:
I'll try to do something with it
UPD: should be better now

DarkSquirrelComes reviewed (changes requested):
- Efficient work with connection pool is the main issue here i see.
-  I'm not sure about config format
- Should we do something with error handling?
- All other small stuff in the comments)))

kolayne committed:
Fixed some issues found during a review

kolayne committed:
Moved `loginForm` inside of `handleLogin`, because it's not used anywhere else

kolayne committed:
Changed some HTTP status codes to submit to the design rules

DarkSquirrelComes commented on src/webhandlers.go:
```diff
@@ -0,0 +1,180 @@
+	if err := os.Chdir("src/templates"); err != nil {
+		return nil, err
+	}
+	defer os.Chdir("../..")
```
WTF?

kolayne committed:
`getAverageColor`: return `{0, 0, 0}` instead of panicking, when called from an empty array

kolayne committed:
Changed api handlers' and some structures' names cases

Exported api handler functions to make it easier to distinguish them from internal functions. Canceled export of internal structs.

Note: this does not affect the behaviour of the program, because the file implements the `main` package

kolayne commented:
Hi! I've made some fixes and decided to rewrite the history, so I had to force-push the branch and the history could change significantly (I say "could" because it did change significantly in my local version, but I don't remember which commits were pushed to the remote branch, so probably there is no big difference for you), sorry about that. Check out the improvements!

VladimirKoryakin reviewed (approved):
I think everything is fine except of probable differences in the database scheme and objects.go if the PR #12 will be merged earlier. But for the current branch master I approve these changes.

kolayne committed:
+ review fixes

`loadHtmlTemplates`: don't change current directory
